velox/connectors/hive/HivePartitionFunction.cpp

```diff
@@ -29,44 +29,106 @@ void hashTyped(
       TypeTraits<kind>::name);
 }
 
-template <>
-void hashTyped<TypeKind::BOOLEAN>(
+template <typename T>
+void abstractHashTyped(
     const DecodedVector& values,
     vector_size_t size,
     bool mix,
-    std::vector<uint32_t>& hashes) {
+    std::vector<uint32_t>& hashes,
+    std::function<uint32_t(const T&)> const& f) {
   for (auto i = 0; i < size; ++i) {
     uint32_t hash;
     if (values.isNullAt(i)) {
       hash = 0;
     } else {
-      hash = values.valueAt<bool>(i) ? 1 : 0;
+      hash = f(values.valueAt<T>(i));
     }
-
     hashes[i] = mix ? hashes[i] * 31 + hash : hash;
   }
 }
 
+template <>
+void hashTyped<TypeKind::BOOLEAN>(
+    const DecodedVector& values,
+    vector_size_t size,
+    bool mix,
+    std::vector<uint32_t>& hashes) {
+  auto f = [](bool value) { return value ? 1 : 0; };
+  abstractHashTyped<bool>(values, size, mix, hashes, f);
+}
+
+template <>
+void hashTyped<TypeKind::TINYINT>(
+    const DecodedVector& values,
+    vector_size_t size,
+    bool mix,
+    std::vector<uint32_t>& hashes) {
+  auto f = [](int8_t value) { return static_cast<uint32_t>(value); };
+  abstractHashTyped<int8_t>(values, size, mix, hashes, f);
+}
+
+template <>
+void hashTyped<TypeKind::SMALLINT>(
+    const DecodedVector& values,
+    vector_size_t size,
+    bool mix,
+    std::vector<uint32_t>& hashes) {
+  auto f = [](int16_t value) { return static_cast<uint32_t>(value); };
+  abstractHashTyped<int16_t>(values, size, mix, hashes, f);
+}
+
+template <>
+void hashTyped<TypeKind::INTEGER>(
+    const DecodedVector& values,
+    vector_size_t size,
+    bool mix,
+    std::vector<uint32_t>& hashes) {
+  auto f = [](int32_t value) { return static_cast<uint32_t>(value); };
+  abstractHashTyped<int32_t>(values, size, mix, hashes, f);
+}
+
+template <>
+void hashTyped<TypeKind::REAL>(
+    const DecodedVector& values,
+    vector_size_t size,
+    bool mix,
+    std::vector<uint32_t>& hashes) {
+  static_assert(sizeof(float) == sizeof(uint32_t));
+  auto f = [](float value) {
+    uint32_t ret;
+    memcpy(&ret, &value, sizeof ret);
```

Comment:
Is it necessary to copy data? Can you use reinterpret_cast instead?

Reply:
In fact I did some searching before doing this change. Seems although `reinterpret_cast` the reference / pointer of `float` to `uint32_t` works, reading the result of it is undefined behavior.
We might be able to use `std::bit_cast` when velox supports using C++20.

Reference:
- https://stackoverflow.com/questions/53995657/is-reinterpret-cast-type-punning-actually-undefined-behavior
- https://timsong-cpp.github.io/cppwp/n4659/basic.lval#8.1
- https://stackoverflow.com/questions/48803363/bitwise-casting-uint32-t-to-float-in-c-c

Reply:
Would you explain a bit more? memcpy is just copying byte-by-byte, it is not type aware, right? Hence, the result would be the same as reinterpret_cast, no?

Reply:
> Would you explain a bit more?
```
float f = 42.0;
uint32_t u = *reinterpret_cast<uint32_t*>(&f);
```
along works in practice in some condition[1]. Underlying, `reinterpret_cast<uint32_t*>(&f)` returns a pointer of type `uint32_t` that actually points to an object of type `float`, this is called **type punning**. As both types are of the same size, we get the exact same result we as do a bit-wise copy.
The issue here is that C++ standard defines "accessing a variable of a given type through a pointer to another type" as undefined behavior, which breaks strict aliasing rule. That is, if type punning is allowed for all the types(not limited in `uint32_t` and `float`), some vital optimize will be prohibited. (I'm not expert in this, so I just accept it as a truth)


> memcpy is just copying byte-by-byte, it is not type aware, right? Hence, the result would be the same as reinterpret_cast, no?

Yes, they both emit the same result. But the difference is by `memcpy`, we copy the actually data byte-by-byte to an object of type `uint32_t`, then when reading it, we are reading an object of type `uint32_t`. This is legal.

[1] To see when it's not working, please compile
```
#include <iostream>

int main()
{
    float f = 42.1;
    uint32_t u = *reinterpret_cast<uint32_t*>(&f);
    std::cout << u << std::endl;
    return 0;
}
```
with
```
g++ -O2 -Wall -Werror
``` 
or more specifically
```
g++ -O2 -Wall -Werror -Wstrict-aliasing
```
you will see the error
```
error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
    6 |     uint32_t u = *reinterpret_cast<uint32_t*>(&f);
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
@usurai Thank you for explaining. @laithsakka Laith, would you take a look? Is there a way to avoid a copy here?

Reply:
@usurai Data copy is expensive, especially one value at a time. I wonder if we could use `values.valueAt<uint32_t>()` instead.

Reply:
> @usurai Data copy is expensive, especially one value at a time. I wonder if we could use `values.valueAt<uint32_t>()` instead.

Oh this is optimal. Pushed new change according to this. Thanks!